Merge SDL3 upstream

capehill opened the pull request.

capehill commented:
Nice, this should be very handy. Some things that might be useful:

a) optimization flag 
b) ggdb flag to provide debug symbols
c) optional altivec: apparently it's not possible to provide a lib that support both altivec and non-altivec machines
d) render/ paths (or did I miss them?)

kas1e commented:
@Capehil
I remember that one of brothers (Thomas?), said that the only debug flag which 100% works fine and can be used on amiga are -gstabs. Other ones may works, but sometime may not

BSzili commented:
> a) optimization flag
> b) ggdb flag to provide debug symbols

How did I miss the optimization and debug flags completely? I'll add these soon. 
> c) optional altivec: apparently it's not possible to provide a lib that support both altivec and non-altivec machines

AltiVec it's a bit tricky, because without -maltivec the intrinsics didn't work at all, so I'll probably disable it completely for the time being.

> d) render/ paths (or did I miss them?)

Could you clarify what you mean?

afxgroup commented:
use gstabs instead of ggdb when generating symbols. because if i remember correctly there was problems reading them with addr2line on os4

capehill commented:
> 
> > d) render/ paths (or did I miss them?)
> 
> Could you clarify what you mean?

This directory should also be built: https://github.com/AmigaPorts/SDL/tree/master/src/render

If I looked correctly, it was missing in the makefile.

BSzili commented:
I got it. I added two commits, let me know if there's anything else I should change.

kas1e commented:
@BSzili
Should't be also src/render/opengles2/ be in makefile as well ? Currently it builded by default inside of SDL2.

BSzili commented:
Beats me. I can't build it, because I don't have the SDK, but someone else can add it to the Makefile and config header.

kas1e commented:
@BSZili
Just in case you will ever need it, there is archive with latest ogles2 sdk:
http://kas1e.mikendezign.com/aos4/ogles2/include_ogles2_2019_02_20.zip

BSzili commented:
Thanks, but is it OK to distribute it publicly? Either way I'll try if I can build the GLES2 support/renderer.

capehill commented on test/Makefile.amigaos4:
```diff
@@ -0,0 +1,282 @@
+
+CC		= ppc-amigaos-gcc
+CFLAGS	= -g -O2 -Wall -g -I../include
+LIBS	= -L.. -lSDL2 -lpthread -lGL -lm
```
I think -lGL and -lm are unnecessary: or did you have some issues without?

BSzili replied:
Some examples need them. Should I link them on a case by case basis?

BSzili replied:
Never mind, the tests built just fine without them. This was probably a remnant of me trying to build them for Windows and forgot about it.

capehill commented on Makefile.amigaos4:
```diff
@@ -13,20 +13,25 @@ TARGET_SHARED  = libSDL2-2.0.so
 	./src/loadso/amigaos4/*.c \
 	./src/render/*.c \
+	./src/power/*.c \
+	./src/power/*.c \
```
Line 31 seems extra

BSzili replied:
I didn't catch this, I'll remove the extra line.

kanjitalk755 committed:
Fixes #8190.
From #7249, reverted the hunks other than #7239.

madebr committed:
cmake: test -Wl,--version-script with minimal version script

Android ndk 26 errors when a symbol in the version script is not defined.

madebr committed:
dynapi: implement SDL_DYNAPI_entry even when building SDL without dynapi support

madebr committed:
Always provide an implementation for all SDL3 symbols

madebr committed:
Add unsupported functions to dynapi

smcv committed:
SDLTest_CompareSurfaces: Log better messages if NULL or different sizes

Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
SDLTest_CompareSurfaces: If surfaces differ, log their formats

Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
SDLTest_CompareSurfaces: Output RGBA values of first differing pixel

Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
pixels: Document the naming convention

Whenever I have to fix something endianness-related, I always get
confused about whether the byte-oriented format that guarantees to put
red in byte 0 is RGBA8888 or RGBA32. (The answer is that it's RGBA32.)

Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
surface: Document the in-memory layout of the pixels

I don't think we ever explicitly said this.

Signed-off-by: Simon McVittie <smcv@collabora.com>

sezero committed:
VisualC/pkg-support/cmake: remove subdir from public header include path

icculus committed:
alsa: More efficient audio thread iteration.

Now we sleep the thread in WaitDevice until ALSA reawakens it because it
needs more data, and we feed it exactly as much as it can take at that
point.

Like the recent PulseAudio changes, this is both more efficient, reliable,
and consistent.

icculus committed:
alsa: Convert `/* */` comments to `//` comments.

icculus committed:
alsa: Simplified PlayDevice and CaptureFromDevice.

icculus committed:
pulseaudio: Use pa_stream_begin_write to avoid an extra buffer copy.

icculus committed:
alsa: Cleaned up remaining debug logging.

slouken committed:
Wait a bit when snd_pcm_avail() returns 0

The hardware might report that it's ready and still need a bit of time to start processing samples

slouken committed:
We don't need to wait a full 10 ms, just delay a bit

icculus committed:
aaudio: Change an int to an SDL_bool.

icculus committed:
aaudio: Deal with device recovery.

Android claims to work with multiple devices, but doesn't actually appear to
(at least, afaict), and it will report tons of devices that all just seem
to play to the current default output, so for now, turn this off and only
expose a default device.

And then, with that default output, attempt to recover on errors by throwing
away the current AAudioStream and building a new one.

This let me plug/unplug a set of headphones from the headphone jack and audio
would switch correctly to the new output.

slouken committed:
Update sample processing bookkeeping when recovering the AAudio audio device

slouken committed:
Use the device audio format for the lowest latency

slouken committed:
Revert "Do a better job of finding default ALSA devices"

This reverts commit d65861f0496a58b2c2642c249f03ac3a08e820c3.

The default device is the one with the name "default"

madebr committed:
ci: disable Intel compilers

smcv committed:
test: Extract SDLTest_ReadSurfacePixel

This is essentially the same as was added in d95d2d70, but with clearer
error handling. It's implemented in a private header file so that it
can be shared with SDL_shape, which also wants this functionality.

Resolves: https://github.com/libsdl-org/SDL/issues/8319
Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
surface: Add a private SDL_ReadSurfacePixel

This shares its implementation with SDLTest_ReadSurfacePixel: the same
code is compiled twice, to get it into the static test library and also
the public shared library.

Signed-off-by: Simon McVittie <smcv@collabora.com>

smcv committed:
shape: Use SDL[Test]_ReadSurfacePixel

This avoids assuming that the pixels are suitably aligned for direct
access, which there's no guarantee that they are; in particular,
3-bytes-per-pixel RGB images are likely to have 3 out of 4 pixels
misaligned. On x86, dereferencing a misaligned pointer does what you
would expect, but on other architectures it's undefined whether it will
work, crash with SIGBUS, or silently give a wrong answer.

Signed-off-by: Simon McVittie <smcv@collabora.com>

Kontrabant committed:
tests: Disable mouse warp test under Wayland

Wayland does not allow warping the mouse pointer, thus these tests are unreliable and should be skipped.

icculus committed:
audio: Commented out a currently-incorrect assert.

Fixes #8326.

Kontrabant committed:
Fix log message spelling

icculus committed:
audio: Destroy the logical audio device before sending DEVICE_REMOVED event.

This prevents catastrophe if someone tries to close the device in an event
filter in response to the event.

Note that this means SDL_GetAudioStreamDevice() for any stream on this
device will return 0 during the event filter!

Fixes #8331.

icculus committed:
audio: Removed unused AllowsArbitraryDeviceNames variable.

icculus committed:
directsound: Cleaned up WaitDevice.

This should retry until GetCurrentPosition succeeds. Otherwise, we would be
going on to the next iteration too soon.

Also generally streamlined the code while I was in here.

icculus committed:
audio: WaitDevice/WaitCaptureDevice now returns a result.

This is an attempt to centralize all the error handling, instead of
implicitly counting on WaitDevice implementations to disconnect the device
to report an error.

icculus committed:
audio: Fixed copy/paste error that was checking wrong variable.

Fixes #8342.

meyraud705 committed:
Do not report gyro/accelerometer if we can't read axes info

icculus committed:
aaudio: Fixed a comment.

madebr committed:
blit_slow: don't read destination pixel when you're going to discard it anyways

madebr committed:
triangle: don't read destination pixel when you're going to discard it anyways

madebr committed:
testdrawchessboard: clean up renderer and window

meyraud705 committed:
Fix assertion in LINUX_JoystickSetSensorsEnabled()

Calling LINUX_JoystickSetSensorsEnabled() after the joystick is unplugged will trigger this assertion.

Etaash-mathamsetty committed:
wayland: Add null check for zenity version string

slouken committed:
Added an example of video decoding with ffmpeg

madebr committed:
cmake: use *_STATIC_* variables when linking to a static ffmpeg

slouken committed:
Allow setting any number of sprites over the video

Default to no sprites over the video

slouken committed:
Make sure we're building with ffmpeg 6.0 or newer

Kontrabant committed:
wayland: Try to skip the Wayland driver if not connecting to or in a Wayland session

When initializing the Wayland driver, check if the application is being started in, or trying to connect to, a Wayland session and skip to another driver if not. If neither WAYLAND_DISPLAY nor XDG_SESSION_TYPE are set, try to start anyway, and if the Wayland library is missing or no Wayland sessions are started, initialization will fail later in the process as it previously did.

This fixes the case where a Wayland session is running on a different VT, but an application wishes to run via KMSDRM on the current VT.

slouken committed:
Added support for 0-copy decode and display using VAAPI and EGL

icculus committed:
add-source-to-projects.pl: Fix adding files in the base src dir.

icculus committed:
hashtable: Added src/SDL_hashtable.[ch].

Reference Issue #7799.

icculus committed:
hashtable: Use Create/Destroy naming, in the SDL3 style.

icculus committed:
audio: Changes to one logical device must update all sibling logical devices.

Fixes #8226.

slouken committed:
Added support for 0-copy decode and display using Apple VideoToolbox

slouken committed:
Added support for 0-copy decode and display using D3D11

FIXME: We need a way to do this that doesn't involve reaching into the D3D11 texture internals

slouken committed:
Added SDL_GetTextureDXGIResource() to get the DXGI resource associated with a render texture.

Also switched the D3D11 and D3D12 renderers to use real NV12 textures for NV12 data.

The combination of these two changes allows us to implement 0-copy video decode and playback for D3D11 in testffmpeg without any access to the renderer internals.

slouken committed:
testffmpeg works with ffmpeg 5.1.3 and newer

sezero committed:
fix testffmpeg.c build.

slouken committed:
Added a "--software" option to testffmpeg

This allows easy performance comparison between hardware and software decoding

icculus committed:
Sync wiki -> header

smcv committed:
audio: Disable NEON sample conversion until test failures are fixed

We need to do this early in the file, so that it will be taken into
account when deciding whether to define NEED_SCALAR_CONVERTER_FALLBACKS
and therefore provide a non-SIMD fallback.

Mitigates: https://github.com/libsdl-org/SDL/issues/8352
Signed-off-by: Simon McVittie <smcv@collabora.com>

madebr committed:
cmake: check ffmpeg capability instead of version

madebr committed:
testffmpeg: use SDL_test to parse arguments and track memory

slouken committed:
Fixed build warning

slouken committed:
Added SDL properties API

Fixes https://github.com/libsdl-org/SDL/issues/7799

slouken committed:
Disable low latency audio by default when using AAudio on Android

Some devices have flat sounding audio when low latency mode is enabled. In particular this was reported on the Samsung SM-T290 running Android 11.

madebr committed:
ci: stop FreeBSD job after 30 minutes

slouken committed:
The sensor and joystick instance ID generator isn't guarded by a lock.

slouken committed:
Fixed warning C4028: formal parameter different from declaration

slouken committed:
Fixed error C2054: expected '(' to follow 'inline'

slouken committed:
Added support for the PowerA Nintendo Switch Nano Wired Controller

icculus committed:
x11: Properly check for XInput2 support before using it.

This specifically fixes a crash in X11_WarpMouseInternal if XInput2 was
missing at runtime, but also cleans up a few other existing checks.

Fixes #8378.

icculus committed:
README-migration.md: Added note about SDL_HasRDTSC removal.

Fixes #8374.

icculus committed:
testaudio: Make program usable without a 3-button mouse.

icculus committed:
audio: Fixed assertions when capture devices have wrong audio formats.

Fixes #8376.

Kontrabant committed:
Use SDL wrapped getenv function

slouken committed:
We don't require the audio system to be initialized for audio format conversion

This is helpful for tools pipelines where audio devices are never used.

slouken committed:
Like mutexes, operations on NULL rwlocks are no-ops

Fixes a crash when performing operations on a NULL rwlock before rwlock implementation is chosen

slouken committed:
Make it clear that the string comparison isn't a boolean check

icculus committed:
pulseaudio: Stop the threaded mainloop before destroying the context.

Otherwise, we might trigger an assertion in libpulse.

Reference Issue #8348.

icculus committed:
audio: Whoops, this stream format change is only for capture devices.

slouken committed:
We need audio converters initialized in SDL_InitAudio()

These are used separately from audio streams, e.g. SDL_OutputAudioThreadIterate(), so they should always be initialized when audio is initialized.

madebr committed:
cmake: optionally install pdb's

slouken committed:
Added properties to various SDL objects

The following objects now have properties that can be user modified:
* SDL_AudioStream
* SDL_Gamepad
* SDL_Joystick
* SDL_RWops
* SDL_Renderer
* SDL_Sensor
* SDL_Surface
* SDL_Texture
* SDL_Window

icculus committed:
audio: Allow audio streams to be created when the subsystem isn't initialized.

You are on your own to destroy them, though!

slouken committed:
Clarified that testffmpeg will resize the window to the video size

slouken committed:
testffmpeg: allow resizing of the video window

slouken committed:
Fixed potential wraparound issue with property IDs

Kontrabant committed:
wayland: Fix primary selection handling when the protocol isn't supported

The primary selection protocol is optional, so the function pointers to the internal Wayland functions should only be set if the protocol is supported. This allows graceful fall-back to the generic SDL implementation in other cases.

Fixes the clipboard tests under Weston.

slouken committed:
testffmpeg: added the ability to specify audio and video codecs

icculus committed:
Sync wiki -> headers

slouken committed:
testffmpeg: added support for YUVA formats using swscale

Fixes https://github.com/libsdl-org/SDL/issues/8377

slouken committed:
testffmpeg: Only enable blending if we're using a texture format that supports it

madebr committed:
testffmpeg: print usage of options to change audio/video codec

kanjitalk755 committed:
Fix macOS build error by #8269

icculus committed:
audio: Never lock a device while holding the device_list_lock.

Fixes various not-necessarily-corner cases ending in deadlock.
